src/smc_sagews/smc_sagews/sage_salvus.py

```diff
@@ -3571,3 +3571,302 @@ def help(*args, **kwds):
+        Returns the list of available Jupyter kernels.
+        '''
+        return os.popen("jupyter kernelspec list").read()
+
+    def _get_doc(self):
+        ds0 = textwrap.dedent(r"""\
+        Use the jupyter command to use any Jupyter kernel that you have installed using from your SageMathCloud worksheet.
+
+            | my_python3 = jupyter("python3")
+
+        After that, begin a sagews cell with the magic command to send statements to the kernel
+
+            | %my_python3
+            | print(42)
+
+        You can set the default mode for all cells in the worksheet. After putting the following
+        in a cell, click the "restart" button, and you have an anaconda worksheet.
+
+            | %auto
+            | a3 = jupyter('anaconda3')
+            | %default_mode a3
+
+        Each magic command connects to its own kernel. So you can have more than
+        one instance of the same kernel type.
+
+            | my_second_python3 = jupyter("python3")
+
+        """)
+        # print("calling JUPYTER._get_doc()")
+        kspec = self.available_kernels()
+        ks2 = string.replace(kspec, "kernels:\n ", "kernels:\n\n|")
+        return ds0 + ks2
+
+    __doc__ = property(_get_doc)
+
+jupyter = JUPYTER()
+
+import jupyter_client
+from Queue import Empty
+from ansi2html import Ansi2HTMLConverter
+import tempfile, sys, re
+import base64
+
+def jkmagic(kernel_name, **kwargs):
+    r"""
+    See docs for jupyter
+    """
+    km, kc = jupyter_client.manager.start_new_kernel(kernel_name = kernel_name)
+
+    kn = kernel_name
+    i_am_a_jupyter_client = True
+
+    debug = kwargs['debug'] if 'debug' in kwargs else False
+
+    def p(*args):
+        if debug:
+            print ' '.join(str(a) for a in args)
+
+    # inline: no header or style tags, useful for full == False
+    # linkify: little gimmik, translates URLs to anchor tags
+    conv = Ansi2HTMLConverter(inline=True, linkify=True)
+
+    salvus.html(conv.convert(""))
+
+    def hout(s, block = True, scroll = False):
+        r"""
+        wrapper for ansi conversion before displaying output
+
+        INPUT:
+
+        -  ``block`` - set false to prevent newlines between output segments
+
+        -  ``scroll`` - set true to put output into scrolling div
+        """
+        # `full = False` or else cell output is huge
+        h = conv.convert(s, full = False)
+        if block:
+            h2 = '<pre style="font-family:monospace;">'+h+'</pre>'
+        else:
+            h2 = '<pre style="display:inline-block;margin-right:-1ch;font-family:monospace;">'+h+'</pre>'
+        if scroll:
+            h2 = '<div style="max-height:320px;width:80%;overflow:auto;">' + h2 + '</div>'
+        salvus.html(h2)
+
+    def big_output(f,*args, **kwargs):
+        r"""
+        Temporarily raise ceilings on output limits (for data uris and docs).
+        Arbitrary new upper limit is 2000000 for both.
+        Restore to prior values after calling the function provided.
+        """
```

Comment:
Please wrap this in a try/finally?  Otherwise if anything goes wrong (or there is an interrupt signal), then the limits aren't restored.